### PR TITLE
more file submission ops

### PR DESCRIPTION
**Pull request by blankdots**
### Related issues

related to: #148 
continuation of #627 

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### Changes Made

- check files have the status `ready` when being read from the submission
- prevent publish if files have in submission have status `added` (added but no metadata object) or `failed` (failed in ingestion, completion, or for any other reason) - in future version we can also check if the submission actually has files
- missing docs about raise exception in files
- update status of a file
  - `/submission/<submission_id>/files` receives a list of files 
  ```
  [
      {
        "accessionId": 1,
        "version": 1,
        "status": "ready".
        "objectId": {
           "accessionId": 4,
           "schema": "study"
        }
      },
      {
        "accessionId": 2,
        "version": 1,
        "status": "ready"
      },
      {
        "accessionId": 3,
        "version": 1,
        "status": "ready"
      }
    ]
  ```
and  transforms it to mongodb query for each file in `update_file_submission` e.g.:
```
{
    "files.$.status": "ready",
    "files.$.version": 1,
    "files.$.objectId": {
      "accessionId": 4,
      "schema": "study"
    }
```

- remove file from submission when flagged as deleted, via `remove_file_submission` which is called when a file is flagged for deletion but can also be called whenever we want to remove all files with a specific accessionId or file path from one or multiple submissions
- operation to assign metadata object to file, also marking a file as ready, done via `/submission/<submission_id>/files` as one or multiple files can be marked as ready - will be used also MQ integration.
- small bug fixes and typo fixes

### Testing

- [x] Needs testing (start an issue or do a follow up PR about it)

### Mentions

this is dependent on #642

### CODE REVIEW

File: metadata_backend/api/operators/file.py [hunk 1]
```diff
@@ -193,18 +202,61 @@ async def read_project_files(self, project_id: str) -> List[Dict]:
...
+        ]
+        try:
+            problematic_files = await self.db_service.do_aggregate("submission", aggregate_query)
+            if len(problematic_files) > 0:
+                reason = (
+                    f"There are a problematic files: {','.join([i['accessionId'] for i in problematic_files])} "
+                    f"in the submission with id: {submission_id}"
+                )
+                LOG.error(reason)
+                raise web.HTTPBadRequest(reason=reason)
```
Comment (blankdots) [inline metadata_backend/api/operators/file.py:238]: not sure this should be bad status, maybe 422, or smth else, any suggestions?

File: metadata_backend/api/operators/file.py [hunk 2]
```diff
@@ -193,18 +202,61 @@ async def read_project_files(self, project_id: str) -> List[Dict]:
...
+                    "version": "$files.version",
+                    "status": "$files.status",
+                }
+            },
+        ]
+        try:
+            problematic_files = await self.db_service.do_aggregate("submission", aggregate_query)
+            if len(problematic_files) > 0:
+                reason = (
+                    f"There are a problematic files: {','.join([i['accessionId'] for i in problematic_files])} "
```
Comment (csc-felipe) [inline metadata_backend/api/operators/file.py:234]: Would the message say that the files are not in use, and should either be used or removed?

Or should we remove unlinked files in the publishing phase? ^^

File: metadata_backend/api/operators/file.py (hunk 1 above)
Comment (csc-felipe) [reply metadata_backend/api/operators/file.py:238]: I think that 400 is good.



_The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing)._

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

Comment (csc-felipe) [reply metadata_backend/api/operators/file.py:238]: Although the examples are quite technical, it seems to me that the other return codes are meant for more specific cases.

I'd consider "forgetting" to either remove the files or attach them to an object as a client error ^^

File: metadata_backend/api/operators/file.py [hunk 3]
```diff
@@ -193,18 +202,61 @@ async def read_project_files(self, project_id: str) -> List[Dict]:
...
+        :raises: HTTPInternalServerError if db operation failed because of connection
+        or other db issue or db aggregate does not return a list
         :returns: List of files specific to a submission
         """
-        # TO_DO: add check for ready status
         aggregate_query = [
             {"$match": {"submissionId": submission_id}},
             {"$unwind": "$files"},
+            # we match only the files that have the status ready
+            {"$match": {"files.status": "ready"}},
```
Comment (csc-felipe) [inline metadata_backend/api/operators/file.py:259]: I'd imagine that we should be able to read submission files in any state that they might be, no?

Should it be a parameter, that if present returns only a matching subset of files, and if not present, returns all files in a submission?

File: metadata_backend/api/operators/file.py (hunk 2 above)
Comment (blankdots) [reply metadata_backend/api/operators/file.py:234]: this is more of a check to log these files, however: i am thinking more also of a retry mechanism if they users keep the files.
So this function might change.

The update (remove or unlink the files) will be done via the submission API, but this needs to be an explicit action.

File: metadata_backend/api/operators/file.py (hunk 3 above)
Comment (blankdots) [reply metadata_backend/api/operators/file.py:259]: ok, I can change that.

**Comment by csc-felipe** [comment]
Looking good

File: metadata_backend/api/operators/file.py (hunk 1 above)
Comment (blankdots) [reply metadata_backend/api/operators/file.py:238]: yes, the plan is to attach them to the error object but for that we might need #641

Comment (blankdots) [reply metadata_backend/api/operators/file.py]: fixed with #642  and https://github.com/CSCfi/metadata-submitter/pull/633/commits/90fef6fa0a2d199967eaea403949fabac1f419d6

File: metadata_backend/api/handlers/submission.py [hunk 4]
```diff
@@ -300,15 +300,16 @@ async def put_submission_path(self, req: Request) -> Response:
...
-
         op = "add"
         if schema in submission:
             op = "replace"
         patch = [
             {"op": op, "path": f"/{schema}", "value": data},
         ]
+
+        submission[schema] = data
+        JSONValidator(submission, "submission").validate
```
Comment (csc-felipe) [inline metadata_backend/api/handlers/submission.py:311]: I don't really see how moving this changed anything. Did I miss something, could you explain, please?

Comment (blankdots) [reply metadata_backend/api/handlers/submission.py:311]: the `if schema in submission` checks if the `schema` is in the `submission`, and having the  `submission[schema] = data`  before the condition check means that the `schema` key will be added to the dict, thus making the op always default to `replace`.

by changing the lines we will not be adding the `schema` to the `submission` dict.

Comment (blankdots) [reply metadata_backend/api/handlers/submission.py:311]: simple example I could think of is:
```
smth = {"smth" : 1}
 
if "cool" in smth:
    print("cool")

smth["cool"] = [1, 2]
```
vs

```
smth = {"smth" : 1}
smth["cool"] = [1, 2]
if "cool" in smth:
    print("cool")
```

Comment (csc-felipe) [reply metadata_backend/api/handlers/submission.py:311]: I got it, thanks 😅

**Comment by blankdots** [comment]
changed base so that mypy fixes are included

File: metadata_backend/database/db_service.py [hunk 5]
```diff
@@ -280,25 +278,43 @@ async def remove(self, collection: str, accession_id: str, data_to_be_removed: U
...
+    ) -> dict:
         """Append data by to object with accessionId in collection.
 
         :param collection: Collection where document should be searched from
         :param accession_id: ID of the object/submission/user to be appended to
         :param data_to_be_addded: str or JSON representing the data that should be
         updated to removed.
         :param upsert: If the document does not exist add it
-        :returns: True if operation was successful
+        :returns: JSON after remove if operaation was successful
```
Comment (csc-jm) [inline metadata_backend/database/db_service.py:317]:
```suggestion
        :returns: JSON after remove if operation was successful
```
🤓

File: metadata_backend/database/db_service.py [hunk 6]
```diff
@@ -264,14 +262,14 @@ async def update_by_key_value(self, collection: str, key: str, value: str, data_
...
-    async def remove(self, collection: str, accession_id: str, data_to_be_removed: Union[str, Dict]) -> bool:
+    async def remove(self, collection: str, accession_id: str, data_to_be_removed: Union[str, Dict]) -> dict:
         """Remove element of object by its accessionId.
 
         :param collection: Collection where document should be searched from
         :param accession_id: ID of the object/submission/user to be updated
         :param data_to_be_removed: str or JSON representing the data that should be
         updated to removed.
-        :returns: True if operation was successful
+        :returns: JSON after remove if operaation was successful
```
Comment (csc-jm) [inline metadata_backend/database/db_service.py:272]:
```suggestion
        :returns: JSON after remove if operation was successful
```
🤓

File: metadata_backend/api/handlers/submission.py [hunk 7]
```diff
@@ -297,19 +297,37 @@ async def put_submission_path(self, req: Request) -> Response:
...
-        JSONValidator(submission, "submission").validate
+        if schema == "files":
+            file_operator = FileOperator(db_client)
+            # we expect to get a list of dict for files
+            # that matches the json schema when added to submission object
+            submission[schema] = data
+            JSONValidator(submission, "submission").validate
+            for file in data:
+                if "accessionId" not in file:
+                    reason = f"Updating {submission_id} failed files requires accessionId."
```
Comment (csc-jm) [inline metadata_backend/api/handlers/submission.py:313]:
```suggestion
                    reason = f"Updating {submission_id} failed. Files require an accessionId."
```
more coherent(?)

File: metadata_backend/api/handlers/submission.py [hunk 8]
```diff
@@ -372,3 +390,28 @@ async def add_submission_files(self, req: Request) -> Response:
         reason = "Request does not contain a list of Objects each with `accessionId` and `version`"
         LOG.error(reason)
         raise web.HTTPBadRequest(reason=reason)
+
+    async def delete_submission_files(self, req: Request) -> Response:
+        """Remove files from a submission.
+
+        Body needs to contain a list of accessionId for files.
+
+        :param req: POST request with metadata schema in the body
```
Comment (csc-jm) [inline metadata_backend/api/handlers/submission.py:399]:
```suggestion
        :param req: DELETE request
```

**Review by csc-jm — COMMENTED** [review]
Found a few things to nitpick but otherwise looks good to me

**Review by csc-jm — APPROVED** [review]

**Review by csc-felipe — APPROVED** [review]
LGTM

Need to create an issue for tests.

**Comment by blankdots** [comment]
> Need to create an issue for tests.

I thought this was it: #629

**Pull request merged**